Document microgrid concepts from an SDK perspective

shsms opened the pull request:
Closes #456

shsms committed:
Add a user guide explaining microgrid concepts from an SDK perspective

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Remove boiler-plate code from LogicalMeter example

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms commented:
Does this one need release notes?

llucax commented on docs/user-guide/microgrid-concepts.md:
```diff
@@ -0,0 +1,128 @@
+
+This refers to a microgrid's connection to the external Grid.  The
+power flowing through this connection can be streamed through
+[`grid_power`][frequenz.sdk.timeseries.logical_meter.LogicalMeter.grid_power].
```
Completely unrelated, but in this context, now that we have a `microgrid.grid` object, I wonder if we shouldn't publish metrics there, like we do with `battery_pool`. We can do it after 1.0 (just adding also `microgrid.grid.power` and deprecating `logical_meter.grid_power`).

shsms replied:
true, that makes sense.

llucax replied:
* #729

llucax commented on docs/user-guide/microgrid-concepts.md:
I would add links to the [glossary](https://github.com/frequenz-floss/frequenz-sdk-python/blob/v1.x.x/docs/user-guide/glossary.md) the first time new terms appear at least: `{{glossary("PV")}}` for example.

shsms replied:
done

llucax commented on docs/user-guide/microgrid-concepts.md:
This might probably go well as the [`frequenz.sdk.microgrid`](https://github.com/frequenz-floss/frequenz-sdk-python/blob/v1.x.x/src/frequenz/sdk/microgrid/__init__.py) module documentation, I would put it there and then include it in `docs/user-guide/microgrid-concepts.md` using:

```md
# Microgrid Concepts

::: frequenz.sdk.microgrid
    options:
        members: []
        show_bases: false
        show_root_heading: false
        show_root_toc_entry: false
        show_source: false
```

Like we did with the actors docs.

shsms replied:
will do this as the last step, so that I can still get syntax highlighting while I make changes.

shsms replied:
done

llucax commented on docs/user-guide/microgrid-concepts.md:
```diff
@@ -0,0 +1,128 @@
+# Microgrid concepts
```
Nitpick as we are normally using this style.
```suggestion
# Microgrid Concepts
```

shsms replied:
done

daniel-zullo-frequenz commented on docs/user-guide/microgrid-concepts.md:
```diff
@@ -0,0 +1,128 @@
+method for fetching the status of individual EV Chargers, including
+state changes like when an EV is connected or disconnected, and a
+[`set_bounds`][frequenz.sdk.timeseries.ev_charger_pool.EVChargerPool.set_bounds]
+method for limit the charge power of individual EV Chargers.
```
Maybe `for limiting` or `to limit` instead?

shsms replied:
done

daniel-zullo-frequenz commented on docs/user-guide/microgrid-concepts.md:
```diff
@@ -0,0 +1,128 @@
+
+## Frequenz SDK Microgrid Model
+
+The SDK aims to provide an abstract model of the microgrid that allows
+for high-level interactions with microgrid components, without having
+to deal with, or even be aware of location specific details like:
```
```suggestion
The SDK aims to provide an abstract model of the microgrid, enabling
high-level interactions with microgrid components, without the need
to handle or be aware of location-specific details such as:
```

shsms replied:
```diff
-The SDK aims to provide an abstract model of the microgrid that allows
+The SDK aims to provide an abstract model of the microgrid, enabling
```
This changes the meaning subtly, from "the model of the microgrid" doing anything to "the sdk" doing something.  So I've changed it to "... that enables"

---

```diff
-without having to deal with, or even be aware of location specific
+without the need to handle or be aware of location-specific
```
I feel like there's a loss of emphasis with this suggestion.  Changing to "... without having to worry about (or even be aware of) location-specific ..."

daniel-zullo-frequenz commented on docs/user-guide/microgrid-concepts.md:
```diff
@@ -0,0 +1,128 @@
+power flowing through this connection can be streamed through
+[`grid_power`][frequenz.sdk.timeseries.logical_meter.LogicalMeter.grid_power].
+
+In locations without a grid connection, this method would still be
+available, and would stream zero values.
```
```suggestion
In locations without a grid connection, this method remains
available and streams zero values.
```

shsms replied:
done, also changed the corresponding sentences for other methods to use the simple present.

daniel-zullo-frequenz commented on docs/user-guide/microgrid-concepts.md:
```diff
@@ -0,0 +1,128 @@
+And total producer power is available through
+[`producer_power`][frequenz.sdk.timeseries.logical_meter.LogicalMeter.producer_power].
+
+And just like the other methods, if components of either of these
+types are not available in a microgrid, the corrensponding methods
+will stream zero values.
```
```suggestion
Similarly to the other methods, if components of either of these
types are unavailable in a microgrid, the corresponding methods
will stream zero values.
```

shsms replied:
Rewrote with "As is the case with the other methods, ..."

daniel-zullo-frequenz reviewed (commented):
I have a few cosmetics to check for

shsms committed:
Link PSC to detailed entry in glossary

Also add some specifics about passive sign convention from the point
of view of the grid connection.

And fix a typo.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Update mkdocs macro `glossary` to take optional link text

This allows linking to a glossary term with custom text for the link.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Improve grammar and tone

Based on review comments

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms reviewed (commented):
Added a commit with the changes

daniel-zullo-frequenz reviewed (commented):
LGTM, I'll leave it to Luca to approve it as he still needs to resolve his comments

shsms committed:
Replace `pathlib.Path.relative_to` with  `os.path.relpath`

The `pathlib` method expects one path to be the sub path of the other.
The `os.path` function is able to find a relative path even if they
diverge, by adding `../`.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Move microgrid docs to the module and reference it in the user guide

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

llucax commented on docs/_scripts/macros.py:
```diff
@@ -45,19 +46,20 @@ def define_env(env: macros.MacrosPlugin) -> None:
         glossary_path = pathlib.Path("user-guide/glossary.md")
-        link_path = glossary_path.relative_to(current_path.parent)
-        return f"[{term}]({link_path}#{_slugify(term)})"
+        link_path = os.path.relpath(glossary_path, current_path.parent)
```
Mmm, interesting. I would add a comment about this because it's not obvious at all so I wouldn't be surprised if someone wants to change it back.

shsms replied:
done

llucax reviewed (dismissed):
I strongly recommend adding a comment but I'm fine to merge as is too.
